for_testing.py

```diff
@@ -0,0 +1,130 @@
+import sys
+import os.path
+
+# from qgis.core import QgsApplication
+# qgishome = 'C:\\Program Files\\QGIS 3.22.9\x07pps\\qgis-ltr'
+# app = QgsApplication([], True)
+# QgsApplication.setPrefixPath(qgishome, True) 
+
+import pandas as pd
+
+from PyQt5.QtCore import *
+from qgis.gui import *
+from qgis.utils import iface
+from qgis import processing
+from qgis.analysis import *
+from qgis.core import *
+from qgis.PyQt.QtCore  import QVariant
+
+import qgis.utils
+
+
+context = QgsProcessingContext() 
+context.setProject(QgsProject.instance())
+
+point_layer = QgsProcessingUtils.mapLayerFromString ("/D:/M.Tech/Fifth Sem/MTP/snr/projection.shp",context)
+
+network_layer = QgsProcessingUtils.mapLayerFromString ("/D:/M.Tech/Fifth Sem/MTP/snr/Transport_Road.shp",context)
+
+# [From this point I am not sure what actually hapening]
+#creating an empty layer
+cLayer = qgis.utils.iface.mapCanvas().currentLayer()
+provider = cLayer.dataProvider()
+#print (provider.encoding(),",", provider.fields(),",", provider.crs())
+
+outputpath = "D:/M.Tech/Fifth Sem/MTP/snr/sinnar_dg.shp"
+
+# SaveVectorOptions contains many settings for the writer process
+# Write to an ESRI Shapefile format dataset using UTF-8 text encoding
+save_options = QgsVectorFileWriter.SaveVectorOptions()
+save_options.driverName = "ESRI Shapefile"
+save_options.fileEncoding = "UTF-8"
+transform_context = QgsProject.instance().transformContext()
+writer = QgsVectorFileWriter.writeAsVectorFormatV3(cLayer,
+                                                 outputpath,
+                                                  transform_context,
+                                                  save_options) #[Not sure if this one is right or wrong]
+
+layer = QgsVectorLayer(outputpath,"rs02082022", "ogr")
+
+if not layer.isValid():
+    print ("Layer failed to load!")
+
+#start editing
+layer.startEditing()
+provider = layer.dataProvider()
+
+#add fields
+provider.addAttributes( [ QgsField("AddressFrom",  QVariant.String)] )
+provider.addAttributes( [ QgsField("AddressTo",  QVariant.String)] )
+layer.commitChanges()
+
+# prepare graph
+vectorLayer  = QgsVectorLayer ("/D:/M.Tech/Fifth Sem/MTP/snr/Transport_Road.shp",'lines')
+director = QgsVectorLayerDirector( vectorLayer, -1, '', '', '', 3 )
+strategy = QgsNetworkDistanceStrategy()
+director.addStrategy( strategy )
+
+builder = QgsGraphBuilder(vectorLayer.crs()) #[ERROR]
+
+# [From this point the code is not working properly.]
+# prepare points
+features = processing.features(point_layer) #[ERROR]

```

Comment:
A better way to open a vector layer and iterate over its features is to use `QgsVectorLayer` API.

```python
        # The layer to be opened is a PostGIS table.
        layer = QgsVectorLayer(self.uri.uri(), "Cadastres", "postgres")
        if not layer.isValid():
            raise Exception(
                "layer cannot be loaded: %s" % self.uri.uri())
        if layer.featureCount() == 0:
            raise Exception( ... )
        for f in layer.getFeatures():

            # Attributes of the feature can be read using attribute name.
            print(f["<column name>"])

            # Attributes of the feature can be updated as follows.
            f["<column name>"] = newvalue

```

Reply:
This one belongs to qgis version 2.x. It is not working in version 3.x

Reply:
The code you paste here I did not get clearly. Can you refer me to anything to understand this?

Reply:
How do count point? 

point_count = point_layer.selectedFeatureCount() is it okay or not?

Reply:
Selected features are the points or routes that are explicitly selected in the map canvas region of the QGIS window. In order to iterate over all features, not just the selected ones, you need the `getFeatures’ API.